C# : Add query to detect SSRF

porcupineyhairs opened the pull request:
This query adds support for detecting Server Side Request Forgery(SSRF) vulnerabilities in C#. This has at least one valid detection in [chhans/Kurs19_Public](https://github.com/chhans/Kurs19_Public). 

I am working on adding a few more sinks. I am new to C#. Can someone tell me which libraries would be good to include here?

johnw545 commented:
Thanks u

tamasvajk reviewed (changes requested):
Thank you for implementing this check. I added some comments inline.
Also, in general, all HTTP request making APIs seem vulnerable to this attack. I think the most common ones are `System.Net.Http.HttpClient`, `System.Net.HttpWebRequest` (` WebRequest.Create()`), and `System.Net.WebClient`; and maybe handling the base classes would also be good, such as `System.Net.Http.HttpMessageInvoker`. Finally, `HttpClient` and `WebClient` allow setting the `BaseAddress` property, so flow into that should also be covered.

porcupineyhairs commented on csharp/ql/src/experimental/CWE-918/RequestForgery.cs:
```diff
@@ -0,0 +1,33 @@
+namespace RequestForgery.Controllers
```
The tests would require stubs. I plan on writing an automated tool to generate them for us. So I am skipping this for now.

tamasvajk replied:
I just learnt that we have a script to do stubbing. I [adjusted it](https://github.com/github/codeql/pull/5453) to use `codeql`. You could give it a try.

porcupineyhairs commented on csharp/ql/src/experimental/CWE-918/RequestForgery.qll:
```diff
@@ -0,0 +1,40 @@
+  override predicate isSource(DataFlow::Node source) { source instanceof Source }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+}
```
I have added a new guard and amended the test my a bit. I think the newer version of the example is better than the previous one.

porcupineyhairs commented on java/ql/src/experimental/CWE-918/RequestForgery.qhelp:
```diff
@@ -1,37 +1,35 @@
-<!DOCTYPE qhelp PUBLIC
-  "-//Semmle//qhelp//EN"
-  "qhelp.dtd">
+<!DOCTYPE qhelp PUBLIC "-//Semmle//qhelp//EN" "qhelp.dtd">
```
removed.

porcupineyhairs commented:
@tamasvajk I have added new sinks for the methods we discussed above. I have also squashed and rebased as the main had diverged quite a bit.

tamasvajk commented on csharp/ql/src/experimental/CWE-918/RequestForgery.qll:
```diff
@@ -0,0 +1,128 @@
+   * to be a guard for Server Side Request Forgery(SSRF) Vulnerabilities.
+   * This guard considers all checks as valid.
+   */
+  private class BaseUriGuard extends SanitizerGuard, MethodCall {
```
I think there could/should be other sanitizers too. The base URL could be checked simply through a string operation.

tamasvajk reviewed (changes requested):
I think the check got a lot better, and could provide value. Here's a [real use case](https://lgtm.com/projects/g/microsoft/BingMapsV8CodeSamples/snapshot/50601d4f62689dc5119e34fb9bf568961c685362/files/Samples/GeoXml/GeoXmlLayer%20-%20Cross%20Domain/GeoXmlProxyService.ashx.cs?sort=name&dir=ASC&mode=heatmap#x38ebd2bdf7282840:1) that's found by it. However at the moment it's quite noisy:
* Most of the HTTP clients accept relative URLs too, and when data flows into a relative URL that should be fine. Maybe an option to overcome this would be to check if there's any assignment to the base address, such as [here](https://lgtm.com/projects/g/saurabhpati/Globomantics/latest/files/Globomantics/Startup.cs?sort=name&dir=ASC&mode=heatmap#L30).
* There are some cases when data flows into the sinks through string concatenation, and not in the base address part of the string: https://lgtm.com/query/1044480058280809211/ or [string interpolation](https://lgtm.com/projects/g/bitwarden/server/snapshot/c90f04bbb4ad375f6aee9ed38d0ec4f10503cae9/files/src/Admin/Controllers/HomeController.cs?sort=name&dir=ASC&mode=heatmap#xbb4ef1825f378321:1).
* The implemented sanitizer is quite specific, there could be other ways to filter too. Let's see if the [LGTM analysis](https://lgtm.com/query/4507035363478054466/) shows any other recurring patterns.

porcupineyhairs commented on csharp/ql/src/experimental/CWE-918/RequestForgery.qll:
```diff
@@ -0,0 +1,202 @@
+    }
+  }
+
+  private class InterpolatedStringSanitizer extends Sanitizer {
```
This check is not working the way I expect it to work. @tamasvajk Can you please take a look?
I am trying to block all flows from the interpolated string except when 
1. It is of the form ``$"http://{`taint`}/blabla/");"``
2. It is something like `$"{`taint`}/blabla/");"`
3. Or when it is used in an unrelated call such as `System.Console.WriteLine($"Getting {taint}");`


However, my current version fails to accommodate for particular two cases which result in 2 false negatives in the result set.

The first occurs in `microsoft/Atlas`. The query fails to alert here as the url parameter is marked sanitized when it is not. I tried changing `InterpolatedStringSanitizer` from a `Sanitizer` to `SanitizerIn` or `SanitizerOut` but niether of them worked. I don't know whats wrong here.

The other two occur in `ddogadkiney/csharp-netcore-sca-02` and `nopSolutions/nopCommerce`. The code base for both of them is quite similar. Seems like both of them are derived from the same code base, The false negative here is due to 
```
$"{(useSsl ? Uri.UriSchemeHttps : Uri.UriSchemeHttp)}{Uri.SchemeDelimiter}{hostHeader.FirstOrDefault()}";
```
Due to the way I have written my Sanitizer for interpolated strings, I can't deal with these type of cases. Considering the fact that the comparison in this case could have been done separately not as part of the interpolated statement, Hard coding the pattern seems bad. Hence, I am left with two alternatives here. Leave these results out of the query or else add a second data flow to consider flow to and from an interpolated string. Considering the fact that adding a flow comes with a heavy performance penalty, I am in for either hard coding the pattern or just using local flow(in this case results would be lost obviously but it's much more performance efficient) or leave the results as is. WDYT?

tamasvajk replied:
I wouldn't care about the second case, where the uri scheme is not hard coded, but flows somehow into the string. It would make the query unnecessarily complex to handle this fully as the delimiter could also flow from a variable or not...

Can you give a link to the false negative in `microsoft/Atlas`?

porcupineyhairs replied:
@tamasvajk https://github.com/microsoft/Atlas/blob/8056fead2b122a711f0e611bd7c0e6b43148c87b/src/Microsoft.Atlas.CommandLine/Execution/WorkflowEngine.cs#L194-L215 
The `_console.WriteLine` on L204 sanitizes the request.url somehow.If you comment out the sanitizer, the alert comes back again.

tamasvajk replied:
Sorry, I still can't repro this. I commented out the sanitizer, but the issue is still not showing up: https://lgtm.com/query/6556373478803825645/

What am I missing?

porcupineyhairs replied:
@tamasvajk The alert is shown on the link you shared. Are you sure you can't see an alert there?

tamasvajk replied:
I see one alert, but it's not the one you linked above in `WorkflowEngine.cs`.

tamasvajk replied:
Sorry, I probably didn't check the path in detail to see that this is actually the issue that you linked above. Coming back to sanitizer: if I don't comment it out, I still get the same result: https://lgtm.com/query/3356943433915439332/ 
So there's no false negative here, is there?

porcupineyhairs replied:
That's interesting. I am trying to run it on my local node and I can still experience the issue. Is the LGTM running on the latest main?

porcupineyhairs replied:
Okay so no I cleared my codeql cache and tried again. it works fine. I can see the alert here. I guess i will mark this as resolved then.

tamasvajk commented on csharp/ql/test/experimental/CWE-918/RequestForgery.cs:
```diff
@@ -0,0 +1,61 @@
+// semmle-extractor-options: /r:System.Text.RegularExpressions.dll /r:HtmlAgilityPack.dll /r:Microsoft.Extensions.Configuration.dll /r:System.Net.Http.dll /r:Microsoft.Extensions.Logging.dll /r:System.Threading.Tasks.dll /r:Microsoft.AspNetCore.Mvc.dll
```
@porcupineyhairs I think you are simply missing `/r:System.Private.Uri.dll` from this line.

porcupineyhairs commented:
@tamasvajk Sorry for the delay. I tried adding that. But now I get the following error message. Can you please help me figure this out? I have rebased this pr to the latest main and added the wip changes temporarily to help debugging.
```
[24399] [ERROR]   Unable to resolve reference 'HtmlAgilityPack.dll'
[24399] [ERROR]   Unable to resolve reference 'Microsoft.Extensions.Configuration.dll'
[24399] [ERROR]   Unable to resolve reference 'Microsoft.Extensions.Logging.dll'
[24399] [ERROR]   Unable to resolve reference 'Microsoft.AspNetCore.Mvc.dll'
[24399] [ERROR]   Compilation error: /home/ydj/work/codeql/csharp/ql/test/experimental/CWE-918/stubs.cs(50,35): error CS0535: 'HtmlNodeCollection' does not implement interface member 'IEnumerable.GetEnumerator()'
[24399] [ERROR]   Compilation error: /home/ydj/work/codeql/csharp/ql/test/experimental/CWE-918/stubs.cs(773,32): error CS0737: 'GroupCollection' does not implement interface member 'IList.this[int]'. 'GroupCollection.this[int]' cannot implement an interface member because it is not public.
[24399] [ERROR]   Compilation error: /home/ydj/work/codeql/csharp/ql/test/experimental/CWE-918/stubs.cs(773,122): error CS0737: 'GroupCollection' does not implement interface member 'IReadOnlyList<Group>.this[int]'. 'GroupCollection.this[int]' cannot implement an interface member because it is not public.
[24399] [ERROR]   Compilation error: /home/ydj/work/codeql/csharp/ql/test/experimental/CWE-918/stubs.cs(773,515): error CS0737: 'GroupCollection' does not implement interface member 'IList<Group>.this[int]'. 'GroupCollection.this[int]' cannot implement an interface member because it is not public.
[24399] [ERROR]   Compilation error: /home/ydj/work/codeql/csharp/ql/test/experimental/CWE-918/stubs.cs(785,12): error CS0111: Type 'GroupCollection' already defines a member called 'this' with the same parameter types
```

tamasvajk commented:
@porcupineyhairs There are multiple issues here:
* `Unable to resolve reference 'HtmlAgilityPack.dll'` and similar issues are caused by the fact that your test is referencing these DLLs, but they can't be found, because the path is not correctly specified. Prepending `../../resources/assemblies/` solves these.
* The errors in the `stubs.cs` are compilation errors. The stub generator seems to have failed on generating a correct stub. In the last days I have worked on the stub generator quite a bit, see here: https://github.com/github/codeql/pull/5664. However, the generator still fails to stub the complete asp.net core public API surface. You could give it a try though, the `csharp/ql/src/Stubs/make_stubs.py` only generates the minimal stub, and I might have fixed the bits that are required for your test case. Another option would be to manually stub the required methods. The latter also shouldn't be too painful, because the main bits of asp.net (not core) are stubbed already, and you could simplify your test file to remove unneeded classes/interfaces, such as `ILogger<HomeController>`, `IConfiguration`. You also don't need `IHttpClientFactory` as you're only using the returned `HttpClient`, which could be manually stubbed with a single `SendAsync` method.

In fact, I gave the latter a quick try, and simplified your test to the below:
```
// semmle-extractor-options: ${testdir}/../../resources/stubs/System.Web.cs /r:System.Threading.Tasks.dll /r:System.Collections.Specialized.dll

using System;
using System.Threading.Tasks;
using System.Web.Mvc;
using System.Net.Http;

namespace SsrfChallenge.Controllers
{
    public class HomeController : Controller
    {
        [HttpPost]
        public async Task<ActionResult> Bad(string url)
        {
            var request = new HttpRequestMessage(HttpMethod.Get, url);

            var client = new HttpClient();
            await client.SendAsync(request);

            return View();
        }
    }
}

// Missing stubs:
namespace System.Net.Http
{
    public class HttpClient
    {
        public async Task SendAsync(HttpRequestMessage request) => throw null;
    }

    public class HttpRequestMessage
    {
        public HttpRequestMessage(HttpMethod method, string requestUri) => throw null;
    }

    public class HttpMethod
    {
        public static readonly HttpMethod Get;
    }
}

```

porcupineyhairs commented:
@tamasvajk Thanks a lot. The tests work fine now. I have pushed the new changes. PTAL.

tamasvajk commented:
> @tamasvajk Thanks a lot. The tests work fine now. I have pushed the new changes. PTAL.

@porcupineyhairs I removed a test case above for simplicity. Could you please add the case with the sanitizer? Also, I'm unsure if you've already seen/covered the comments in https://github.com/github/codeql/pull/5110#pullrequestreview-616274705.

porcupineyhairs commented:
@tamasvajk I have made the changes now. I have also started a [new run](https://lgtm.com/query/5563564306781972350/) on the 61 projects which threw an alert. There are 12 independent projects. All of them appear to be TP. PTAL.

porcupineyhairs commented:
@tamasvajk I have squashed and rebased to the latest main while this is pending review.

tamasvajk commented on csharp/ql/src/semmle/code/csharp/exprs/Expr.qll:
````diff
@@ -869,7 +869,7 @@ class NameOfExpr extends Expr, @nameof_expr {
  * ```
  */
-class InterpolatedStringExpr extends Expr, @interpolated_string_expr {
+class  InterpolatedStringExpr extends Expr, @interpolated_string_expr {
````
A CI check is failing due to this unneeded formatting change.

tamasvajk reviewed (changes requested):
I've restarted the LGTM analysis: https://lgtm.com/query/5526907158523374183/

porcupineyhairs commented:
@tamasvajk Changes done!

tamasvajk commented:
@hvitved Could you please also have a final look at this PR?

hvitved commented on csharp/ql/src/experimental/CWE-918/RequestForgery.qll:
```diff
@@ -0,0 +1,240 @@
+   */
+  abstract private class Sink extends DataFlow::ExprNode { }
+
+  /**
+   * A data flow Barrier for server side request forgery vulnerabilities.
+   */
+  abstract private class Barrier extends DataFlow::ExprNode { }
+
+  /**
+   * A data flow node which does not allow taint to flow into it
+   * when analysing server side request forgery vulnerabilities.
+   */
+  abstract private class BarrierIn extends DataFlow::ExprNode { }
+
+  /**
+   * A data flow node which does not allow taint to flow outside it
+   * when analysing server side request forgery vulnerabilities.
+   */
+  abstract private class BarrierOut extends DataFlow::ExprNode { }
```
These three classes do not appear to be used, so remove.

hvitved commented on csharp/ql/src/experimental/CWE-918/RequestForgery.qll:
```diff
@@ -0,0 +1,240 @@
+    StringStartsWithBarrierGuard() { this.getTarget().hasQualifiedName("System.String.StartsWith") }
+
+    override predicate checks(Expr e, AbstractValue v) {
+      (e = this.getQualifier() or this.getArgument(0) = e) and
```
Why is the 0th argument guarded?

porcupineyhairs replied:
added a comment to explain this better.

porcupineyhairs commented on csharp/ql/src/experimental/CWE-918/RequestForgery.qll:
```diff
@@ -0,0 +1,240 @@
+    }
+
+    override predicate checks(Expr e, AbstractValue v) {
+      (e = this.getArgument(0) or e = this.getQualifier()) and
```
I have added a comment to explain this better now.

porcupineyhairs commented on csharp/ql/src/experimental/CWE-918/RequestForgery.qll:
```diff
@@ -0,0 +1,240 @@
+
+    override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+
+    override predicate isAdditionalFlowStep(DataFlow::Node prev, DataFlow::Node succ) {
```
We don't want to alert on cases where the taint is the second operand such as in ```url = "http://example.com"+taint```.

porcupineyhairs commented:
@hvitved Sorry for the delay is responding to your comments. I have made some changes now. PTAL.

hvitved commented:
The tests are failing:
```
ERROR: Could not resolve type Barrier (/home/runner/work/semmle-code/semmle-code/ql/csharp/ql/src/experimental/CWE-918/RequestForgery.qll:58,79-86)
```

porcupineyhairs commented:
@tamasvajk Changes done! PTAL.

tamasvajk commented:
@porcupineyhairs I think you might have forgotten to push your changes. The last commit in this PR is from July, and since then @hvitved also provided some feedback. Could you please double check?

porcupineyhairs commented:
@tamasvajk @hvitved Yes, indeed, that was the case, I didn't push the changes here properly. I have rebased to the latest main and forced pushed the changes now. PTAL.

hvitved commented:
The `RequestForgery.qlref` test is still failing:
```
--- expected
+++ actual
@@ -3,5 +3,6 @@
 nodes
 | RequestForgery.cs:14:52:14:54 | url : String | semmle.label | url : String |
 | RequestForgery.cs:16:66:16:68 | access to parameter url | semmle.label | access to parameter url |
+subpaths
 #select
 | RequestForgery.cs:16:66:16:68 | access to parameter url | RequestForgery.cs:14:52:14:54 | url : String | RequestForgery.cs:16:66:16:68 | access to parameter url | $@ flows to here and is used in a method of WebClient. | RequestForgery.cs:14:52:14:54 | url | User-provided value |
```

porcupineyhairs commented:
@hvitved It seems like the `subpath` section in the output was added recently to CodeQL. My CLI version was from earlier and hence, it didn't include it. I have added the new section again. I can confirm the test run clear on my local node now.
